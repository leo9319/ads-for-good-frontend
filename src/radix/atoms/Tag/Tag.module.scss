@import '../../../styles/colours';

.tagComp {
  @mixin button-mode(
    $color,
    $background-color,
    $hover-background-color,
    $active-background-color,
    $active-color: #fff,
    $border: none,
    $shadow: 0px 0px 0px 4px #1049a94d
  ) {
    background-color: $background-color;
    box-shadow: none;
    border: $border;
    color: $color;
    font-family: var(--typography-default-font-family);
    font-weight: 400;
    font-size: 14px;

    &:hover {
      background-color: $hover-background-color;
    }

    &:focus:not(:hover) {
      box-shadow: $shadow;
    }

    &:active {
      background-color: $active-background-color;
      color: $active-color;
    }
  }

  .default {
    @include button-mode(
      $color-neutral-950,
      $color-neutral-50,
      $color-neutral-100,
      $color-blue-500
    );
  }

  .emergency {
    @include button-mode(
      $color-white,
      $color-red-600,
      $color-red-700,
      $color-red-600
    );
  }

  .tagText {
    font-weight: 400;
    color: $color-neutral-950;
    font-size: var(--typography-font-size-1);
    line-height: var(--typography-line-height-1);
  }
}
