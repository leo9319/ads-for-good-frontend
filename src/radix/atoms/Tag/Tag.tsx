import React, { ReactElement } from 'react';
import { Avatar, Badge } from '@radix-ui/themes';
import { CrossCircledIcon } from '@radix-ui/react-icons';

import styles from './Tag.module.scss';
import Text from '../Text';

/**
 * @interface
 * The Tag props type of {@link Tag}
 */
export interface TagProps {
  /**
   * Pass the text label to display for the tag.
   * @defaultValue `Text`
   * @property
   */
  text?: string;

  /**
   * Pass the mode to change the tag's appearance.
   * @defaultValue `primary`
   * @property
   */
  mode?: string;

  /**
   * Enable the element's click event if it is true.
   * @property
   */
  isClickable?: boolean;

  /**
   * Display the avatar (circled) image if it is true and iconUrl is available;
   * otherwise, the fallback value will be visible.
   * @property
   */
  showAvatar?: boolean;

  /**
   * Display the icon on the left side of the text if it is true.
   * @property
   */
  showIcon?: boolean;

  /**
   * Pass a valid image URL or filepath to display the avatar image.
   * @property
   */
  iconUrl?: string;
}

/**
 * Tag component to display the tag with required information such as text, icon, and avatar.
 *
 * @group Atoms
 * @category Component
 *
 * @param {TagProps} props - The Tag props
 * @returns {ReactElement} - Tag Component
 *
 * @example
 * ```
 * <Tag text="Text" showAvatar={true} mode="secondary" />
 * ```
 */
export const Tag = ({
  isClickable,
  showIcon,
  showAvatar,
  iconUrl,
  text = 'Text',
  mode = 'default',
}: TagProps): ReactElement => {
  return (
    <div
      className={styles.tagComp}
      style={{ pointerEvents: isClickable ? 'auto' : 'none' }}
    >
      <Badge
        style={{
          paddingInline: showAvatar ? '4px 12px' : '8px 12px',
          borderRadius: '20px',
          cursor: isClickable ? 'pointer' : 'unset',
          width: 'auto',
          maxWidth: '100%',
          height: '32px',
          paddingBlock: 0,
        }}
        className={styles[mode]}
        tabIndex={0}
      >
        {showIcon && <CrossCircledIcon />}
        {showAvatar && (
          <Avatar
            radius="full"
            src={iconUrl}
            fallback={text.slice(0, 2) || 'AB'}
            size="1"
          />
        )}
        <Text truncate className={styles.tagText}>
          {text}
        </Text>
      </Badge>
    </div>
  );
};

export default Tag;
