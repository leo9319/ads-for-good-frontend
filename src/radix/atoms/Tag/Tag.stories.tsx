import React from 'react';
import type { Meta, StoryObj } from '@storybook/react';
import Tag from '@radix-styles/atoms/Tag';
import { Theme } from '@radix-ui/themes';

// More on how to set up stories at: https://storybook.js.org/docs/writing-stories#default-exportF
const meta = {
  title: 'Library/Atoms/Tag',
  component: Tag,
  decorators: [
    Story => (
      <Theme>
        <Story />
      </Theme>
    ),
  ],
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/configure/story-layout
    layout: 'centered',
  },
  tags: ['autodocs'],
  // More on argTypes: https://storybook.js.org/docs/api/argtypes
  argTypes: {
    text: {
      control: { type: 'text' },
    },
    mode: {
      control: { type: 'select' },
      options: ['default', 'emergency'],
    },
    isClickable: {
      options: [true, false],
      control: { type: 'boolean' },
    },
    showIcon: {
      options: [true, false],
      control: { type: 'boolean' },
    },
    showAvatar: {
      options: [true, false],
      control: { type: 'boolean' },
    },
    iconUrl: {
      control: { type: 'text' },
    },
  },
  // more on args: https://storybook.js.org/docs/writing-stories/args#args-object
  args: {
    mode: 'default',
    isClickable: true,
  },
} satisfies Meta<typeof Tag>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/writing-stories/args

export const TextOnly: Story = {
  args: {
    text: 'Text only',
    showAvatar: false,
    showIcon: false,
    iconUrl: '',
  },
};

export const IconPlusText: Story = {
  args: {
    showIcon: true,
    text: 'Icon with Text',
    showAvatar: false,
    iconUrl: '',
  },
};

export const AvatarPlusText: Story = {
  args: {
    showAvatar: true,
    text: 'Avatar',
    iconUrl:
      'https://images.unsplash.com/photo-1607346256330-dee7af15f7c5?&w=64&h=64&dpr=2&q=70&crop=focalpoint&fp-x=0.67&fp-y=0.5&fp-z=1.4&fit=crop',
    showIcon: false,
  },
};

export const Emergency: Story = {
  args: {
    mode: 'emergency',
    showIcon: true,
    text: 'Emergency',
    showAvatar: false,
    iconUrl: '',
  },
};
