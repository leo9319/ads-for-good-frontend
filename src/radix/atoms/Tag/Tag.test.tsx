import React from 'react';
import { render } from '@testing-library/react';
import Tag from './Tag';

describe('Tag', () => {
  it('Should render successfully', () => {
    const { baseElement } = render(<Tag text="Text" />);
    expect(baseElement).toMatchSnapshot();
  });
});
