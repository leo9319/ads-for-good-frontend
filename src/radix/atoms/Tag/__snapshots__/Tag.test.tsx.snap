// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag Should render successfully 1`] = `
<body>
  <div>
    <div
      class="tagComp"
      style="pointer-events: none;"
    >
      <span
        class="rt-reset rt-Badge rt-r-size-1 rt-variant-soft default"
        data-accent-color=""
        style="padding-inline: 8px 12px; border-radius: 20px; cursor: unset; width: auto; max-width: 100%; height: 32px; padding-block: 0;"
        tabindex="0"
      >
        <span
          class="rt-Text rt-truncate mocked-class-name typography-font-family-default tagText"
          id=""
        >
          Text
        </span>
      </span>
    </div>
  </div>
</body>
`;
